java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -333,15 +337,21 @@ abstract class AbstractArcHost(
             )
             particleContext.handles[handleName] = handle
             if (handleConnection.mode.canRead) {
+                // Once all of the readable handles for this particle have received their
+                // [StorageEvent.READY] notification, we need to call [Particle.onReady].
+                particleContext.expectReady(handle)
+            }
+        }
+
+        // Once we've expected-ready from all of our readable handles, we can bind the
+        // ParticleContext's lifecycle-notification method to the handles' storage events.
+        particleContext.handles.forEach { (name, handle) ->
+            if (spec.handles[name]?.mode?.canRead == true) {
                 // Particles with readable handles need to be notified for storage events against
                 // those handles, but a direct connection is difficult in the current architecture.
                 // Instead, we'll use the [ParticleContext] instance to manage the particle
                 // lifecycle APIs and thread events from the StorageProxy up via a callback.
                 handle.registerForStorageEvents { particleContext.notify(it, handle) }
```

Comment:
How does this lead to onReady being called twice? I see this
```
if (awaitingReady.remove(handle) && awaitingReady.isEmpty()) {
                    particleState = ParticleState.Running
                    particle.onReady()
                }
```
? It seems like this can only be called once in that code.

Reply:
Imagine two handles for the particle, A and B.

With the old mechanism, this sequence would be possible:

1. register for storage events on A
1. expect ready for A
1. A becomes ready
1. Since A was the only item in `awaitingReady`, we trigger `particle.onReady()`
1. register for storage events on B
1. expect ready for B
1. B becomes ready
1. Since B was the only item in `awaitingReady`, we trigger `particle.onReady()`.

Reply:
In other words - if the handle becomes ready really quickly, before the other handles are added to the `awaitingReady` list, you can get bonus `onReady` callbacks.

Reply:
Ah, ok, thanks for the explanation.

Reply:
@jasonwyatt I'm not sure your example is actually possible. In the old code, the startup sequence was:
- `startParticle` is called for all particle specs:
  - creates handles and preps their proxies for sync (moving them to `READY_TO_SYNC`)
  - `registerForStorageEvents` adds a notify callback to the proxy, _which can't be synced yet_
  - `expectReady` adds to the `awaitingReady` set
- `entityHandleManager.initiateProxySync` is called to kick off sync requests for all the arc's proxies (moving them to `AWAITING_SYNC`)

Since the proxies will only start to sync at that last call, there shouldn't be any way for a handle to become ready before all particles and their handles have been set up.

I'm moving a lot of this code around to improve the lifecycle state machine, so I would like to understand the problem here to avoid re-introducing bugs.

Reply:
I was definitely seeing more than one onReady come through. It maybe it's possible that the bonus events are caused when two handles share the same storage proxy (and thus, share a listener on the store), but one handle is part of a particle which has others... and the second handle is just a standalone one - who initiated the sync request. This could make the race possible.